internal/pkg/db/redis/data.go

```diff
@@ -203,6 +217,122 @@ func (c *Client) DeleteEventById(id string) (err error) {
 	return nil
 }
 
+// DeleteEventsByDevice Delete events and readings associated with the specified deviceID
+func (c *Client) DeleteEventsByDevice(deviceId string) (int, error) {
+	err := c.DeleteReadingsByDevice(deviceId)
+	if err != nil {
+		return 0, err
+	}
+
+	conn := c.Pool.Get()
+	defer conn.Close()
+
+	ids, err := redis.Strings(conn.Do("ZRANGE", db.EventsCollection+":device:"+deviceId, 0, -1))
+	if err != nil {
+		return 0, err
+	}
+
+	err = conn.Send("MULTI")
+	if err != nil {
+		return 0, err
+	}
+
+	for _, id := range ids {
+		err = conn.Send("RENAME", id, db.DeletedEventsCollection+":"+id)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	err = conn.Send("EXEC")
+	deleteEventsChannel <- deviceId
+
+	return len(ids), nil
+}
+
+// AsyncDeleteEvents Handles the deletion of device events asynchronously. This function is expected to be running in
+// a go-routine and works with the "DeleteEventsByDevice" function for better performance.
+func (c *Client) AsyncDeleteEvents(loggingClient logger.LoggingClient) {
+	loggingClient.Debug("Starting background event deletion process")
+	for {
+		select {
+		case device, ok := <-deleteEventsChannel:
+			if ok {
+				loggingClient.Debug("Deleting event data for device: " + device)
+				startTime := time.Now()
+				c.deleteRenamedEvents(device, loggingClient)
+				loggingClient.Debug(fmt.Sprintf("Deleted events for device: '%s', elapsed time: %s", device, time.Since(startTime)))
+			}
```

Comment:
See the comment about closing the channel above. If a read happens when the channel is closed, then ok will be false and this method should explicitly return -- which will, in turn, free up resources assigned to the goroutine.